File: python/lesson 15/longest_increasing_subsequence.py
# https://leetcode.com/problems/longest-increasing-subsequence/

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        piles = []    
        for num in nums:
            toInsert = self.searchInsert(piles, num)
            if toInsert == len(piles):
                piles.append(num)
            else:
                piles[toInsert] = num
        return len(piles)

    def searchInsert(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if target < nums[mid]:
                hi = mid - 1
            elif target > nums[mid]:
                lo = mid + 1
            else:
                return mid
        return lo

File: csharp/lesson 15/LongestIncreasingSubsequence.cs
// https://leetcode.com/problems/longest-increasing-subsequence/

public class Solution
{
    public int LengthOfLIS(int[] nums)
    {
        var piles = new List<int>();
        
        foreach (var num in nums)
        {
            var toInsert = SearchInsert(piles, num);
            if (toInsert == piles.Count)
            {
                piles.Add(num);
            } 
            else
            {
                piles[toInsert] = num;
            }
        }
        
        return piles.Count;
    }
    
    private int SearchInsert(List<int> piles, int target)
    {
        int lo = 0;
        int hi = piles.Count - 1;
        
        while (lo <= hi)
        {
            var mid = lo + (hi - lo) / 2;
            var value = piles[mid];            
            if (value == target)
            {
                return mid;
            } 
            else if (value < target) 
            {
                lo = mid + 1;
            }
            else
            {
                hi = mid - 1;
            }
        }
        
        return lo;
    }
}

